pkg/cosigner/image_signer.go

```diff
@@ -0,0 +1,101 @@
+package cosigner
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+
+	"github.com/buildpacks/lifecycle"
+	"github.com/sigstore/cosign/cmd/cosign/cli"
+)
+
+type SignFunc func(ctx context.Context, ko cli.KeyOpts, annotations map[string]interface{}, imageRef string, certPath string, upload bool, payloadPath string, force bool, recursive bool) error
+
+type ImageSigner struct {
+	Logger   *log.Logger
+	signFunc SignFunc
+}
+
+const (
+	cosignRepositoryEnv       = "COSIGN_REPOSITORY"
+	cosignDockerMediaTypesEnv = "COSIGN_DOCKER_MEDIA_TYPES"
+)
+
+var (
+	secretLocation = "/var/build-secrets/cosign"
+)
+
+func NewImageSigner(logger *log.Logger, signFunc SignFunc) *ImageSigner {
+	return &ImageSigner{
+		Logger:   logger,
+		signFunc: signFunc,
+	}
+}
+
+func (s *ImageSigner) Sign(ctx context.Context, report lifecycle.ExportReport, annotations map[string]interface{}, cosignRepositories map[string]interface{}, cosignDockerMediaTypes map[string]interface{}) error {
+	cosignSecrets, err := findCosignSecrets()
+	if err != nil {
+		s.Logger.Printf("no keys found for cosign signing: %v\n", err)
+		return nil
+	}
+
+	if len(cosignSecrets) == 0 {
+		s.Logger.Println("no keys found for cosign signing")
+		return nil
+	}
+
+	refImage := report.Image.Tags[0]
+
+	for _, cosignSecret := range cosignSecrets {
+		cosignKeyFile := fmt.Sprintf("%s/%s/cosign.key", secretLocation, cosignSecret)
+		cosignPasswordFile := fmt.Sprintf("%s/%s/cosign.password", secretLocation, cosignSecret)
+
+		ko := cli.KeyOpts{KeyRef: cosignKeyFile, PassFunc: func(bool) ([]byte, error) {
+			content, err := ioutil.ReadFile(cosignPasswordFile)
+			if err != nil {
+				return []byte(""), nil
```

Comment:
should this return the err?

Reply:
It shouldnt error. "cosign.password" is an optional file or field in the secret. I will add a comment saying so here

Reply:
We could, however, output a warning message just in case the password is missing but the key requires it (that would be a configuration error that can be hard to spot without logs).

Reply:
maybe something like this just to be safe:

```
content, err := ioutil.ReadFile(cosignPasswordFile)
  if err == nil {
    return content, nil
  } else if errors.Is(err, os.ErrNotExist) {
    return []byte(""), nil
  } else {
    return []byte(""), err
}
```

Reply:
So the file could not exist or the file could be empty. When it is empty, the password would be an empty string based on the file content.